Sources/SourceKitLSP/Rename.swift

```diff
@@ -284,78 +291,438 @@ private extension RenameLocation.Usage {
+          )
+        }
+      }
+      return try await swiftToClangTranslationTask!.value
+    case is ClangLanguageServerShim:
+      return definitionName
+    default:
+      throw ResponseError.unknown("Cannot rename symbol because it is defined in an unknown language")
+    }
+  }
+
+  /// Get the name of the symbol as it should be used to perform a rename in Swift files.
+  ///
+  /// This runs the clang importer to translates Objective-C selectors to Swift functions.
+  ///
+  /// - Note: The symbol translation from clang to Swift requires a Swift location at which the clang symbol is
+  ///   (probably so that sourcekitd can access the necessary AST context and get the clang importer to translate the
+  ///   name). But the output does not actually matter on the specific position and URI. We can thus cache the result.
+  ///   Ideally, sourcekitd wouldn't need a position and URI to translate a clang name to Swift.
+  func swiftName(
+    at position: Position,
+    in snapshot: DocumentSnapshot,
+    languageService: SwiftLanguageServer
+  ) async throws -> String {
+    switch definitionLanguageService {
+    case is SwiftLanguageServer:
+      return definitionName
+    case is ClangLanguageServerShim:
+      if clangToSwiftTranslationTask == nil {
+        clangToSwiftTranslationTask = Task {
+          try await languageService.translateClangNameToSwift(
+            at: position,
+            in: snapshot,
+            isObjectiveCSelector: isObjectiveCSelector,
+            name: definitionName
+          )
+        }
+      }
+      return try await clangToSwiftTranslationTask!.value
+    default:
+      throw ResponseError.unknown("Cannot rename symbol because it is defined in an unknown language")
+    }
+  }
+}
+
 // MARK: - SourceKitServer
 
 extension SourceKitServer {
+  private func getTranslatableName(
+    forUsr usr: String,
+    workspace: Workspace,
+    index: IndexStoreDB
+  ) async throws -> (name: TranslatableName, language: Language)? {
+    guard let definitionSymbol = index.occurrences(ofUSR: usr, roles: [.definition]).only else {
+      logger.error("Multiple or no definitions for \(usr) found")
+      return nil
+    }
+    let definitionLanguage: Language =
+      switch definitionSymbol.symbol.language {
+      case .c: .c
+      case .cxx: .cpp
+      case .objc: .objective_c
+      case .swift: .swift
+      }
+    let definitionDocumentUri = DocumentURI(URL(fileURLWithPath: definitionSymbol.location.path))
+
+    let isObjectiveCSelector =
+      definitionLanguage == .objective_c
+      && (definitionSymbol.symbol.kind == .instanceMethod || definitionSymbol.symbol.kind == .classMethod)
+
+    guard
+      let nativeLanguageService = await self.languageService(
+        for: definitionDocumentUri,
+        definitionLanguage,
+        in: workspace
+      )
+    else {
+      logger.fault("Failed to get language service for the document defining \(usr)")
+      return nil
+    }
```

Comment:
I was going to suggest passing in the language service and then using it if the primary file is the same as the definition file, but it's pretty cheap. One thing I did notice though was that we currently log which toolchain we're using for every `languageService` call, which seems like it would be pretty spammy for this request.

Reply:
It’s only being logged on the `info` level, which isn’t included in the logs unless explicitly requested. The logs will be somewhat spammy because we log a `getSyntacticRenameRanges` request for every file as well. And who knows, it might actually useful to see which toolchain we are using for all these files. So, I’d just keep it for now.

Reply:
You're the one looking at them right now, so 👍 Can always change in the future if it's too spammy.